seed/seedwriter/fetcher.go

```diff
@@ -0,0 +1,87 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package seedwriter
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/asserts"
+)
+
+// SeedAssertionFetcher is a Fetcher which is designed to help with the fetching
+// of assertions during seeding. It keeps track of assertions fetched, and allows
+// for retrieving them at any point in time during seeding. It wraps around the
+// asserts.{SequenceFormingFetcher,Fetcher} interfaces to allow for flexible
+// retrieval of assertions.
+type SeedAssertionFetcher interface {
+	Fetch(ref *asserts.Ref) error
```

Comment:
Would it make sense to do something like:
```go
--- a/seed/seedwriter/fetcher.go
+++ b/seed/seedwriter/fetcher.go
@@ -31,9 +31,8 @@ import (
 // asserts.{SequenceFormingFetcher,Fetcher} interfaces to allow for flexible
 // retrieval of assertions.
 type SeedAssertionFetcher interface {
-       Fetch(ref *asserts.Ref) error
-       FetchSequence(seq *asserts.AtSequence) error
-       Save(asserts.Assertion) error
+       asserts.SequenceFormingFetcher
+
        Refs() []*asserts.Ref
        ResetRefs()
 }
```
here? Or do we want to make sure this and the fetchers are "de-coupled"?

Reply:
Because of the optionality of the fetcher implementing the entire interface of SequenceFormingFetcher, it makes it more confusing that it directly refers the interface (in my opinion naturally), instead being explicit about the interface makes more sense.